core/src/main/java/org/hisp/dhis/android/core/trackedentity/TrackedEntityAttributeReservedValueManager.kt

```diff
@@ -0,0 +1,407 @@
+        val attributeOrgunit = if (isOrgunitDependent(pattern)) organisationUnitUid else null
+
+        return store.popOne(attributeUid, attributeOrgunit)?.value() ?: throw D2Error.builder()
+            .errorCode(D2ErrorCode.NO_RESERVED_VALUES).errorDescription("There are no reserved values")
+            .errorComponent(D2ErrorComponent.Database).build()
+    }
+
+    /**
+     * @param attributeUid           Attribute uid
+     * @param numberOfValuesToFillUp An optional maximum number of values to reserve
+     * @see .downloadReservedValues
+     */
+    fun blockingDownloadReservedValues(
+        attributeUid: String,
+        numberOfValuesToFillUp: Int?
+    ) {
+        runBlocking { downloadReservedValues(attributeUid, numberOfValuesToFillUp) }
+    }
+
+    /**
+     * Download of TrackedEntityInstance reserved values. The number of reserved values is filled up to the
+     * numberOfValuesToFillUp. If not defined, it defaults to [.FILL_UP_TO].
+     * <br></br><br></br>
+     * The download is only triggered for this attribute passed as parameter.
+     * <br></br><br></br>
+     * If the attribute pattern is dependent on OrganisationUnit code (that is, it contains "ORG_UNIT_CODE"), it
+     * reserves values for each orgunit assigned to the programs with this attribute. It applies the limit
+     * per orgunit. Otherwise the limit is applied per attribute.
+     *
+     * @param attributeUid           Attribute uid
+     * @param numberOfValuesToFillUp An optional maximum number of values to reserve
+     * @return An Observable that notifies about the progress.
+     */
+
+    fun downloadReservedValues(
+        attributeUid: String,
+        numberOfValuesToFillUp: Int?
+    ): Observable<D2Progress> {
+        return downloadReservedValuesFlow(attributeUid, numberOfValuesToFillUp).asObservable()
+    }
+
+    private fun downloadReservedValuesFlow(
+        attributeUid: String,
+        numberOfValuesToFillUp: Int?
+    ) = flow<D2Progress> {
+        downloadValuesForOrgUnits(attributeUid, numberOfValuesToFillUp).collect()
+    }
+
+    /**
+     * @param numberOfValuesToFillUp An optional maximum number of values to reserve
+     * @see .downloadAllReservedValues
+     */
+    fun blockingDownloadAllReservedValues(numberOfValuesToFillUp: Int?) {
+        runBlocking {
+            downloadAllReservedValues(numberOfValuesToFillUp)
+        }
+    }
+
+    /**
+     * Downloads reserved values for all the trackedEntityAttributeValues of type "generated", that is, it applies
+     * [.downloadReservedValues] for every generated attribute.
+     *
+     * @param numberOfValuesToFillUp An optional maximum number of values to reserve
+     * @return An Observable that notifies about the progress.
+     */
+
+    fun downloadAllReservedValues(numberOfValuesToFillUp: Int?): Observable<D2Progress> {
+        return rxObservable { downloadAllReservedValuesFlow(numberOfValuesToFillUp).collect() }
```

Comment:
I think the use of `rxObservable` is not correct here. It emits with `send()`, something like 
`rxObservable { downloadAllReservedValuesFlow(numberOfValuesToFillUp).collect { send(it) } }`

It is easier to write it as:
`downloadAllReservedValuesFlow(numberOfValuesToFillUp).asObservable()`

Reply:
That was because of the `downloadAllReservedValuesFlow` design with a `return` value; I should've refactored it to just ```    private fun downloadAllReservedValuesFlow(
        numberOfValuesToFillUp: Int?
    ) = flow {
        val flows = generatedAttributes.map { attribute ->
            downloadValuesForOrgUnits(attribute.uid(), numberOfValuesToFillUp)
        }

        emitAll(flows.merge())
    }```